### PR TITLE
Initial setup for Docker

### PR DESCRIPTION
**Pull request by alihussain03**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,6 @@
 # ConnectedCircularity
+
+### Docker Compose
```
Comment (bunjovis) [inline README.md:3]: Maybe add note that you should add .env file for each service before.

File: cc-backend/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+# Materiaalitori service setup.
+
+# Start by copying the server stub to this directory.
+
+
+# First line is provided for your convenience.
+# This line fetches the Node Docker image with tag "16"
+# from Docker Hub.
+# We use version 16 as the mongodb dependency has requirements
+FROM node:latest
```
Comment (bunjovis) [inline cc-backend/Dockerfile]: The comment above does not reflect this line. Should we use some fixed version of node? Also we don't have mongodb

File: cc-db-service/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# Materiaalitori service setup.
+
+# Start by copying the server stub to this directory.
+
+
+# First line is provided for your convenience.
+# This line fetches the Node Docker image with tag "16"
+# from Docker Hub.
+# We use version 16 as the mongodb dependency has requirements
+FROM node:latest
```
Comment (bunjovis) [inline cc-db-service/Dockerfile:9]: The comment above does not reflect this line. Should we use some fixed version of node? Also we don't have mongodb

File: cc-mt-service/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+# Materiaalitori service setup.
+
+# Start by copying the server stub to this directory.
+
+
+# First line is provided for your convenience.
+# This line fetches the Node Docker image with tag "16"
+# from Docker Hub.
+# We use version 16 as the mongodb dependency has requirements
+FROM node:latest
```
Comment (bunjovis) [inline cc-mt-service/Dockerfile:9]: The comment above does not reflect this line. Should we use some fixed version of node? Also we don't have mongodb

File: cc-pk-service/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+# Materiaalitori service setup.
+
+# Start by copying the server stub to this directory.
+
+
+# First line is provided for your convenience.
+# This line fetches the Node Docker image with tag "16"
+# from Docker Hub.
+# We use version 16 as the mongodb dependency has requirements
+FROM node:latest
```
Comment (bunjovis) [inline cc-pk-service/Dockerfile]: The comment above does not reflect this line. Should we use some fixed version of node? Also we don't have mongodb

File: cc-backend/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+WORKDIR /usr/src/app
+
+# Install app dependencies
+# A wildcard is used to ensure both package.json AND package-lock.json are copied
+# where available (npm@5+)
+COPY package*.json ./
+
+RUN npm install
+
+RUN npm install -g typescript
```
Comment (bunjovis) [inline cc-backend/Dockerfile:22]: Is this necessary? Isn't it already installed with npm install?

File: cc-backend/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+RUN npm install express
```
Comment (bunjovis) [inline cc-backend/Dockerfile]: Is this necessary? Isn't it already installed with npm install?

File: cc-db-service/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+WORKDIR /usr/src/app
+
+# Install app dependencies
+# A wildcard is used to ensure both package.json AND package-lock.json are copied
+# where available (npm@5+)
+COPY package*.json ./
+
+RUN npm install
+
+RUN npm install -g typescript
```
Comment (bunjovis) [inline cc-db-service/Dockerfile:23]: Is this necessary? Isn't it already installed with npm install?

File: cc-db-service/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+RUN npm install express
```
Comment (bunjovis) [inline cc-db-service/Dockerfile]: Is this necessary? Isn't it already installed with npm install?

File: cc-mt-service/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+WORKDIR /usr/src/app
+
+# Install app dependencies
+# A wildcard is used to ensure both package.json AND package-lock.json are copied
+# where available (npm@5+)
+COPY package*.json ./
+
+RUN npm install
+
+RUN npm install -g typescript
```
Comment (bunjovis) [inline cc-mt-service/Dockerfile:23]: Is this necessary? Isn't it already installed with npm install?

File: cc-mt-service/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+RUN npm install express
```
Comment (bunjovis) [inline cc-mt-service/Dockerfile]: Is this necessary? Isn't it already installed with npm install?

File: cc-pk-service/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+WORKDIR /usr/src/app
+
+# Install app dependencies
+# A wildcard is used to ensure both package.json AND package-lock.json are copied
+# where available (npm@5+)
+COPY package*.json ./
+
+RUN npm install
+
+RUN npm install -g typescript
```
Comment (bunjovis) [inline cc-pk-service/Dockerfile:20]: Is this necessary? Isn't it already installed with npm install?

File: cc-pk-service/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+RUN npm install express
```
Comment (bunjovis) [inline cc-pk-service/Dockerfile]: Is this necessary? Isn't it already installed with npm install?

File: cc-backend/readme.md [hunk 14]
```diff
@@ -21,3 +21,7 @@
 - nodeJS 16.14.0
 - typescript 4.8.3
 
+### Docker Guide
```
Comment (bunjovis) [inline cc-backend/readme.md:24]: Maybe add note that you should add .env file before

File: cc-db-service/README.md [hunk 15]
```diff
@@ -16,3 +16,8 @@ Additional scripts:
 
 1. Run `npm run check` to check the file for TypeScript errors/warnings with eslint
 2. Run `npm run dev` to run nodemon
+
+### Docker Guide
```
Comment (bunjovis) [inline cc-db-service/README.md:22]: Maybe add note that you should add .env file before

File: cc-mt-service/README.md [hunk 16]
```diff
@@ -7,3 +7,8 @@
    - ```node dist/app.js```. Runs the created JavaScript file with Node
 
 Additional script: Run ```npm run check``` to check the file for TypeScript errors/warnings with eslint
+
+### Docker Guide
```
Comment (bunjovis) [inline cc-mt-service/README.md:12]: Maybe add note that you should add .env file before

File: cc-pk-service/README.md [hunk 17]
```diff
@@ -13,3 +13,8 @@
 Additional scripts: 
 1. Run ```npm run check``` to check the file for TypeScript errors/warnings with eslint
 2. Run ```npm run dev``` to run nodemon
+
+### Docker Guide
```
Comment (bunjovis) [inline cc-pk-service/README.md:17]: Maybe add note that you should add .env file before

File: docker-compose.yml [hunk 18]
```diff
@@ -0,0 +1,25 @@
+# This is the docker-compose-yml file that must be able to start all project
+
+version: "3.9"
+
+services:
+
+  cc-backend:
+   build: cc-backend/ # use the Dockerfile for Materiaalitori_service
```
Comment (hannakortetjarvi) [inline docker-compose.yml:8]: comment mentions materiaalitori_service, change that to backend

File: docker-compose.yml [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+services:
+
+  cc-backend:
+   build: cc-backend/ # use the Dockerfile for Materiaalitori_service 
+   ports:
+     - '3000:3000'
+
+  cc-db-service:
+   build: cc-db-service/ # use the Dockerfile for Materiaalitori_service
```
Comment (hannakortetjarvi) [inline docker-compose.yml:13]: comment mentions materiaalitori_service, change that to db service

File: docker-compose.yml [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+   ports:
+     - '4001:4001'
+
+  cc-mt-service:
+   build: cc-mt-service/ # use the Dockerfile for Materiaalitori_service 
+   ports:
+     - '8000:8000'
+  
+  cc-pk-service:
+    build: cc-pk-service/ # use the Dockerfile for Materiaalitori_service
```
Comment (hannakortetjarvi) [inline docker-compose.yml:23]: comment mentions materiaalitori_service, change that to Purkukartoitus_service

File: cc-backend/Dockerfile [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+
+
+COPY . .
+
+EXPOSE 3000
+
+CMD ["npm", "run", "dev" ]
```
Comment (rvsape) [inline cc-backend/Dockerfile:28]: the cc-backend might need to run the build command before running

**Comment by rvsape** [comment]
the port mappings in the docker-compose file are fine at this stage of development. later on we could just expose the port for cc-ui, so that it is the only entry point for the system
also the port configurations should be able to read from an .env file

File: README.md (hunk 1 above)
Comment (alihussain03) [reply README.md:3]: Done

File: cc-backend/Dockerfile (hunk 2 above)
Comment (alihussain03) [reply cc-backend/Dockerfile]: Updated

File: cc-db-service/Dockerfile (hunk 3 above)
Comment (alihussain03) [reply cc-db-service/Dockerfile:9]: Updated

File: cc-mt-service/Dockerfile (hunk 4 above)
Comment (alihussain03) [reply cc-mt-service/Dockerfile:9]: updated

File: cc-pk-service/Dockerfile (hunk 5 above)
Comment (alihussain03) [reply cc-pk-service/Dockerfile]: updated

File: cc-backend/readme.md (hunk 14 above)
Comment (alihussain03) [reply cc-backend/readme.md:24]: done

File: cc-db-service/README.md (hunk 15 above)
Comment (alihussain03) [reply cc-db-service/README.md:22]: done

File: cc-mt-service/README.md (hunk 16 above)
Comment (alihussain03) [reply cc-mt-service/README.md:12]: done

File: cc-pk-service/README.md (hunk 17 above)
Comment (alihussain03) [reply cc-pk-service/README.md:17]: done

File: docker-compose.yml (hunk 18 above)
Comment (alihussain03) [reply docker-compose.yml:8]: done

File: docker-compose.yml (hunk 19 above)
Comment (alihussain03) [reply docker-compose.yml:13]: done

File: docker-compose.yml (hunk 20 above)
Comment (alihussain03) [reply docker-compose.yml:23]: done

File: cc-backend/Dockerfile (hunk 21 above)
Comment (alihussain03) [reply cc-backend/Dockerfile:28]: done

File: cc-pk-service/Dockerfile (hunk 13 above)
Comment (alihussain03) [reply cc-pk-service/Dockerfile]: updated. express installation

Comment (alihussain03) [reply cc-pk-service/Dockerfile]: i am not sure about typescript part because i faced issue and install manually typescript.

**Review by bunjovis — CHANGES_REQUESTED** [review]
I guess the typescript install lines are necessary then? 
Some fixes still needed:
cc-mt-service is still missing instructions about the .env file. cc-ui is missing dockerfile. cc-pk-service dockerfile is missing the .env file creation.

**Comment by alihussain03** [comment]
1. can you please elaborate more what is means by "cc-mt-service is still missing instructions about the .env file" ? .env file is creating in docker. do i need to update the README or something else.
2. i was facing issue in cc-ui about ui port. i post that in teams but did not get the response yet. i will update again
3. Regarding the cc-pk-service dockerfile, i need to discuss because we need to provide the credentials in .env file then how to handle this thing ? we cannot provide credentials in docker file. maybe need to look for some solution like read data from some other source.

**Comment by bunjovis** [comment]
> 1. can you please elaborate more what is means by "cc-mt-service is still missing instructions about the .env file" ? .env file is creating in docker. do i need to update the README or something else.
> 
>     2. i was facing issue in cc-ui about ui port. i post that in teams but did not get the response yet. i will update again
> 
>     3. Regarding the cc-pk-service dockerfile, i need to discuss because we need to provide the credentials in .env file then how to handle this thing ? we cannot provide credentials in docker file. maybe need to look for some solution like read data from some other source.

1. yes, so the readme is like other services. 
2. good
3. well that is true. the answer here is then: no need to create the .env file in dockerfile. user just has to make sure they have read README and created the .env file, otherwise it will not work.

**Review by bunjovis — DISMISSED** [review]

**Review by rvsape — CHANGES_REQUESTED** [review]
Add .dockerignore - file in the root of the whole project and ignore node_modules

**Review by bunjovis — APPROVED** [review]

**Review by rvsape — APPROVED** [review]

**Pull request merged**